service/state/operations.go

```diff
@@ -0,0 +1,216 @@
+	return func(txn *badger.Txn) error {
+		var value []byte
+		err := retrieve(Encode(prefixIndexCommit, commit), &value)(txn)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve commit height: %w", err)
+		}
+
+		*height = binary.BigEndian.Uint64(value)
+
+		return nil
+	}
+}
+
+func RetrieveBlockHeight(blockID []byte, height *uint64) func(txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		var value []byte
+		err := retrieve(Encode(prefixIndexHeight, blockID), &value)(txn)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve block height: %w", err)
+		}
+
+		*height = binary.BigEndian.Uint64(value)
+
+		return nil
+	}
+}
+
+func RetrieveCommit(height uint64, commit *[]byte) func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return retrieveCompressed(Encode(prefixIndexHeight, height), &commit)(txn)
+	}
+}
+
+func RetrieveHeader(height uint64, header *flow.Header) func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return retrieveCompressed(Encode(prefixDataHeader, height), &header)(txn)
+	}
+}
+
+func SetLastCommit(commit []byte) func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return txn.Set(Encode(prefixLastCommit), commit)
+	}
+}
+
+func SetCommitHeight(commit []byte, height uint64) func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return  txn.Set(Encode(prefixIndexCommit, commit), Encode(height))
+	}
+}
+
+func SetHeightCommit(height uint64, commit []byte) func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return  txn.Set(Encode(prefixIndexHeight, height), commit)
+	}
+}
+
+func SetBlockHeight(blockID []byte, height uint64) func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return  txn.Set(Encode(prefixIndexBlock, blockID), Encode(height))
+	}
+}
+
+func SetHeader(height uint64, header *flow.Header) func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return setCompressed(Encode(prefixDataHeader, height), header)(txn)
+	}
+}
+
+func SetDeltas(height uint64, change dps.Change)func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return setCompressed(Encode(prefixDataDelta, change.Path, height), change.Payload)(txn)
+	}
+}
+
+func SetEvents(height, typeHash uint64, events []flow.Event) func (txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		return setCompressed(Encode(prefixDataEvents, height, typeHash), events)(txn)
+	}
+}
+
+func retrieve(key []byte, value *[]byte) func(txn *badger.Txn) error {
+	return func(txn *badger.Txn) error {
+		item, err := txn.Get(key)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve value: %w", err)
+		}
+
+		val, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("unable to copy value: %w", err)
+		}
+
+		*value = val
+
+		return nil
+	}
+}
+
+func retrieveCompressed(key []byte, value interface{}) func(txn *badger.Txn) error {
```

Comment:
I really don't think we need both. There is no reason _not_ to compress every value. I didn't do it before because the gain for small values is minimal and it would have created a lot of extra code. Now that we have it all factored into a single re-usable function, we can just compress every value instead.